### ISSUE TITLE
The error messages could contain more server-produced information

### ISSUE THREAD

**Issue by onyxmaster** [elastic/elastic-transport-net#63]
Hi!

Thanks for building this library, it's a great piece of software!

I would like to suggest an improvement, targeting quality of life for the users of this library, increasing diagnostic value of messages that are produced by the library.
When an invalid request is presented to the Elasticsearch server, it fails with the HTTP error 400. The library reports the reason and enriches the message with additional details, like server error, for example: "Type: search_phase_execution_exception Reason: "all shards failed"". Unfortunately, the [`Error.ToString`](https://github.com/elastic/elastic-transport-net/blob/38d8f548028eee7b46ddd05451fea0b9bf2a0c53/src/Elastic.Transport/Products/Elasticsearch/Failures/Error.cs#L31) only takes into account the `CausedBy` field when constructing the message, ignoring the `RootCause`. This leads to server error string representation to miss crucial context in such cases, while the `RootCause` contains the exact cause of the error, for example `Reason` being set to "failed to create query: For input string: "Test"", which would make the user understand the issue much faster, especially when post-mortem debugging, looking only at a log. I understand that the same messages can be sourced, for example, from the Elasticsearch own logs, but I believe that having this information right in the exception message is much more convenient.

So, I would like to suggest adding the root cause (maybe only first of the items) to the implementation of the `Error.ToString()`.

Thanks for considering this proposal.

**Comment by stevejgordon** [comment]
Hi, @onyxmaster.

The `ToString` override here is generally intended for a quick and relatively short error summary. The `RootCause` property contains a collection of other potential root causes. It's exposed as a property so a consumer can access each of those, and each `ErrorCause` then includes a `ToString` override.

There's no one correct answer here as, for some consumers, it's a convenient shortcut for logging. For others, they need to access the properties they care about to access more detail. I feel it's better not to try and provide everything on `ToString` by default, as it adds overhead. You could write an extension method to format your own message from the `Error` properties.

It sounds like your suggestion is specific to a version of the [.NET client](https://github.com/elastic/elasticsearch-net) rather than this lower-level transport library. If so, it's better suited in that repository. To help me understand better, can you provide a specific example, with code, so I can better understand you scenario? If you're using a client, which version?

**Comment by onyxmaster** [comment]
Thank you for the prompt reply.

Yes, the issue is actually about the .NET client library, it's just that this repository contained the code I thought is the target for change, so I decided to file the issue here.

Let me describe our scenario: we're using `IElasticClient.Search` method, when the query is invalid (our latest case had wrong type of data, we serialized enum name instead of an integer), this leads to an HTTP error 400, which is thrown with a following message: `Elasticsearch.Net.ElasticsearchClientException: The remote server returned an error: (400) Bad Request. Call: Status code 400 from: POST /.../_search?typed_keys=true. ServerError: Type: search_phase_execution_exception Reason: "all shards failed"`. Well, it's good that I know that it's the query that is the problem, but I don't know the specifics :( The instance of an `ElasticsearchClientException` that gets thrown does not seem to contain the `IElasticsearchResponse` or an `Error` instance, so it I don't think I can interrogate it for the root cause, it is already lost at the time I can catch the exception. If I'm mistaken, I would be glad to know it =)

Of course I can run the thing under the debugger and find out why it failed, but it has its own limitations, it would be more convenient if the exception itself contained the server-reported root cause.

I've only seen only one kind of root cause and it was quite short ("failed to create query: For input string: "Test""), but I understand that it could be quite long, so I agree with you that adding it to the `Error.ToString` might not be the best idea in all of the cases.

**Comment by stevejgordon** [comment]
Thanks for the extra information, @onyxmaster. I presume you are configuring the client settings with `ThrowExceptions`. If so, then the `Error` is not available on the exception. That might be an enhancement we can make but I'd have to look at the flow in depth. Generally, most users don't configure the `ThrowExceptions(true)` option and in that case, you can access the `ServerError` on the response. A workaround here would be to prefer that, if you require the extra information.

**Comment by onyxmaster** [comment]
Yes, we enable `ThrowExceptions(true)` as a part of "fail early, fail fast" approach without need for manual checking.
Of course I understand that you need to take a closer look at scope of the change.
Thanks for the advice on the potential workaround. Since we've got IElasticClient hidden behind adapters it should not be too hard for us to switch from exceptions being thrown by the library to the ones being thrown by us. But I still think it might be beneficial for other users to have this information in the exception that is thrown by the library, so I'll hope for that outcome =)
